cirq/work/collector.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Optional, Any, TypeVar, Generic, NamedTuple, Iterable, Union, \
+    List
+
+import numpy as np
+
+from cirq import circuits, study
+from cirq.work import sampler, work_pool
+
+TResult = TypeVar('TResult')
+TKey = TypeVar('TKey')
+
+
+class CircuitSampleJob(NamedTuple('CircuitSampleJobBase', [
+    ('circuit', circuits.Circuit),
+    ('repetitions', int),
+    ('key', TKey)
+]), Generic[TKey]):
+    """Describes a sampling task."""
+    pass
+
+
+class SampleCollector(Generic[TKey, TResult], metaclass=abc.ABCMeta):
+    """An interface for concurrently collecting sample data."""
+
+    @abc.abstractmethod
+    def next_job(self) -> Union[None,
+                                CircuitSampleJob[TKey],
+                                Iterable[CircuitSampleJob[TKey]]]:
+        """Called by driving code when more sampling can be started.
+
+        Returns:
+            A CircuitSampleJob describing the circuit to sample, how many
+            samples to take, and a key value that can be used in the
+            `on_job_result` method to recognize which job this is.
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_job_result(self,
+                      job: CircuitSampleJob,
+                      result: study.TrialResult) -> None:
+        """Called by driving code when sample results have become available.
+
+        The results should be incorporated into the collector's state.
+        """
+        pass
+
+    def result(self) -> TResult:
+        """Produces the collector's final statistic."""
+        pass
+
+
+async def async_collect_samples(collector: SampleCollector[Any, TResult],
+                                sampler: sampler.Sampler,
+                                *,
+                                min_concurrent_jobs: int = 1,
+                                max_total_samples: Optional[int] = None) -> TResult:
+    """Concurrently collects samples from a simulator or hardware.
+
```

Comment:
Think this should include an example of how to use this.  Many won't be aware of "await" and async in python so especially important here.